### DISCUSSION TITLE
[DISCUSSION]combine으로 비동기처리를 할때 loadingview 관리를 하는 방법에 대한 의견

### DISCUSSION THREAD

**Discussion by kimscastle** [Team-Plu/Plu-iOS#60] [Q&A] [answered]
현재 온보딩에서 닉네임을 서버에 등록하는 비동기처리 과정에서 loadingview의 active상태를 publish해야하는 상황입니다

제가 생각한 방식은 두가지 입니다

## 첫번째 방식
첫번째방식은 버튼을 눌렀을때 비동기처리의 결과는 void로 반환하지만 내부적으로 loadingstate를 받고 보내주는 passthroughsubject에 start와 end를 보내고 그 subject를 vc에서 받아서 state마다 loadingview의 animation을 켜고 끄고 해주는 방식입니다

```swift
final class OnboardingViewModelImpl: OnboardingViewModel, NicknameCheck {
    
    enum OnboardingFlowType {
        case backButtonTapped, signInButtonTapped
    }

    ...생략...

    var navigationSubject = PassthroughSubject<OnboardingFlowType, Never>()
    let signInStatePublisher = PassthroughSubject<LoadingState, Never>()
    var cancelBag = Set<AnyCancellable>()
    
    ...생략...
    
    func transform(input: OnboardingInput) -> OnboardingOutput {
        
        ...생략...
        
        // 이 부분 입니다
        input.singInButtonTapped
            .flatMap { nickname -> AnyPublisher<Void, Never> in
                return Future<Void, Error> { promise in
                    Task {
                        do {
                            self.signInStatePublisher.send(.start)
                            try await Task.sleep(nanoseconds: 100_000_000_0)
                            try await self.nickNameManager.registerUser(nickName: nickname)
                            self.signInStatePublisher.send(.end)
                            self.navigationSubject.send(.signInButtonTapped)
                            promise(.success(()))
                        } catch {
                            promise(.failure(error))
                        }
                    }
                }
                .catch { _ in
                    Just(())
                }
                .eraseToAnyPublisher()
            }
            .eraseToAnyPublisher()
            .sink {}
            .store(in: &cancelBag)
  
        return OnboardingOutput(nickNameResultPublisher: nickNameResultPublisher, signInStatePublisher: signInStatePublisher)
    }
}
```
- 만약에 LoadingState를 따로 publish하는 publisher가 있다면 vc입장에선 loadingview의 시작시점에대한 아무런 관심이없어져서 VC는 버튼 action의 시점만 전달하면됩니다
- 제가 조금 찜찜한 부분은 이렇게 하면 버튼 action을 따로 sink를 해줘야한다는 점입니다(publish를 위한 단순 빈 sink)
- 그리고 두번째로 찜찜한 부분은 해당 Task내에서 너무 많은 일을 하고있는게 아닌가 라는 생각이 듭니다(state보내주고, flow보내주고, 비동기결과값 보내주고)

## 두번째 방식
두번째 방식은 애초에 비동기처리의 결과로 loadingview state를 publish해주는 방식입니다
이 경우에는 promise로 값을 보내주기에 promise가 불리면 completion block이 불려 star의 경우엔 의미가 없어집니다
> vc에서 버튼이 눌리는 순간 loadingview를 start시키고 비동기처리가 끝나면 promise로부터 .end를 받아 loadingview가 멈춤

```swift
func bindInput() {
    
    ...생략...
    
    self.signInButton.tapPublisher
        .sink { [weak self] in
            guard let text = self?.nickNameTextField.text else { return }
            // 버튼을 누르면 loadingview 시작
            self?.loadingView.startAnimating()
            self?.singInButtonTapped.send(text)
            self?.signInButton.isUserInteractionEnabled = false
        }
        .store(in: &cancelBag)
}
```
- start에 대한 시점을 vc에서 명시를 해줘야합니다

그리고 future쪽 로직이 아래와 같이 변경됩니다
```swift
let signInStatePublisher = input.singInButtonTapped
    .flatMap { nickname -> AnyPublisher<LoadingState, Never> in
        return Future<LoadingState, Error> { promise in
            Task {
                do {
                    try await Task.sleep(nanoseconds: 100_000_000_0)
                    try await self.nickNameManager.registerUser(nickName: nickname)
                    self.navigationSubject.send(.signInButtonTapped)
                    promise(.success(.end))
                } catch {
                    promise(.failure(error))
                }
            }
        }
        .catch { _ in
            Just(.error)
        }
        .eraseToAnyPublisher()
    }
    .eraseToAnyPublisher()
```
- 이 방식으로 하면 하나의 로직이 주어진 일만 하는 느낌이 드는데 loadingview에대한 로직이 vc와 vm에 나눠져서 위치하고있다는 느낌이 들어조금 찜찜합니다...
- 그리고 사실 promise위치때문에 push된다음에 end가 publish되어서 stop되는게 큰 의미가있을까싶기도하고요(push후에 end가 호출되긴합니다 하지만 UI가 바뀌고난뒤에 stop되는게 의미가 있을지는 모르겠네요)

---

두 방식이 제가 생각한 loadingview에대한 state 처리 하는 방식인데 둘다 똑같이 동작을합니다
만약에 loadingview의 state에 대한 처리를 한곳에서만 하고싶다하면 위의방식이 나아보이지만 조금더 로직을 분리하고싶다고하면 아래방식이 좋아보이긴합니다

그래서 더좋은방식이나 혹은 두 방식중 어떤 방식이 더 나아보이는지 궁금해서 discussion을 만들어봤습니다
답변부탁드립니다!!
아마도 이제 loadingview의 state를 두분다 publish하긴해야할거라 여기서 통일해서 mvvm하면 좋을거같아요
@ffalswo2 @cchanmi

**Comment by ffalswo2** [comment, marked as answer]
저는 후자의 방식이 조금 더 맞다고 생각됩니다.

`loadingView`도 결국 `UIView`이고, `loadingView`를 start시키고 hide시키는건 **UI 업데이트하는 쪽에 속한다**고 생각됩니다. 따라서 UI를 단순히 업데이트 하는 로직(start, hide) 자체는 UI와 UI로직의 책임을 가지고 있는 `ViewController`에 있는 것이 더 자연스럽고 있어도 이상하지 않다고 생각합니다. 그리고 `ViewController`에게 start시킬지 hide시킬지를 명령하는 주체는 결국 `ViewModel`이기에 해당 `loadingView`의 state를 따로 저장하진 않지만, **loadingView의 state또한 ViewModel이 결정한다고 봐도 무방**할 것 같습니다. 

저는 이게 loadingView의 로직이 나뉘어졌다고 생각이 들진 않고 오히려 객체가 각자의 책임에 충실하게 해야할 일을 하는 것 같이 느껴집니다.

전자의 경우 loadingView를 start, hide하는 UI업데이트까지의 역할을 수행하고 있어 오히려 ViewController의 책임을 침범하는 느낌이 드는 것 같습니다.

**Reply by kimscastle** [reply]
UI를 시작하는건 VC의 담당이고 이후 비동기처리의 결과로 loading을 hide하라고 VC에 명령하는건 VM의 역할이라고 생각 하면 후자의 방식이 조금더 맞는 방식은 맞긴하겠네요

해당 방식으로 구현해보도록하겠습니다~

근데 그렇게되면 애초에 LoadingState라는 enum이 필요가 없을수도있겠네요
저는 LoadingState라는 enum에 start, end, error이렇게 세가지 case를 넣어놨었는데 사실 네트워크 통신의 결과로 error가 발생하더라도 hide가 되어야하는건 똑같고 error로인해 hide된경우엔 추가적인 action이 필요하겠지만 그건 내부적으로 catch내부에서 error처리하는 곳으로 어떤 값을 보내줘서 처리를 할테니 후자의 경우엔 start가 필요없을거고 .error를 보낸다는거자체가 vc가 어떤방식으로든 error를 알게되는거같아서 그냥 bool값으로 보내는것도 한번 고려를 해봐야겠네요

해당 방식이라면 애초에 start는 vc에서 할거고 hide할 시점만 알려주면되서 enum이 필요가 없을거같긴하네요
물론 true false가 hide해야한다는 직관적인 명령이 아니라 좀 애매한데
하나의 case만으로 명확하게 hide라는 case를 보낼지 아니면 단순하게 bool로 할지 고민이되네요

**Reply by ffalswo2** [reply]
저는 아래와 같이 성공적으로 성공한 경우에 end(success가 조금 더 자연스러운 네이밍일 것 같긴하지만 Result타입자체의 success안에 또 success가 따라붙는게 어색해서 이와 같은 네이밍으로 했습니다.) 그리고 실패한 경우에 연관값에 에러 메시지를 담아서 일단 보내는 형태로 enum을 보내서 구분하긴 했습니다. 두 경우 모두 Loading을 끝내줘야하는 시점인 건 동일하지만 성공과 실패에 대한 구분은 필요할 것 같아서 이렇게 했습니다. 요부분은 한번 팀내에서 이야기해보고 통일해봐도 좋을듯 싶네요

```swift
enum LoginState {
    case end
    case error(message: String)
}
```
